ProductSearch.ipynb

```diff
@@ -0,0 +1,298 @@
+    "\n",
+    "Below is a quick description of the input to the DocRetrieval constructor:\n",
+    "* dense_input_dimension is the dimension of the output of our embedding model (128 with our preloaded model)\n",
+    "\n",
+    "* num_tables and hashes_per_table are hyperparameters of the index\n",
+    "  * increasing num_tables increases the accuracy of the model at the cost of speed and memory (a good value is the preset 16)\n",
+    "  \n",
+    "  * hashes_per_table has a sweet spot in accuracy at around log_2(average document size)\n",
+    "  \n",
+    "* centroids is a numpy array that represents precomputed centroids for the embedding space; we've already calculated and stored these for you! We've precomputed 2^18 of them, but really you just need the number of centroids to be a small fraction of the total documents you intend to add. A simple heuristic to get less centroids is to select a random subsample of them, which we do below (we select the number of centroids equal to 10% of the number of documents)."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "34678e9f",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "embedding_model = DocSearchModel()\n",
+    "\n",
+    "all_centroids = embedding_model.getCentroids().copy()\n",
+    "np.random.shuffle(all_centroids)\n",
+    "reduced_centroids = all_centroids[: max(len(dataset) // 10, 1)]\n",
+    "\n",
+    "our_index = DocRetrieval(\n",
+    "    dense_input_dimension=128,\n",
+    "    num_tables=16,\n",
+    "    hashes_per_table=6,\n",
+    "    centroids=reduced_centroids,\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "b6936e83",
+   "metadata": {},
+   "source": [
+    "# Step 3: Populating the index\n",
+    "We'll now simply loop through each document in our generated dataset, embed the document text, and\n",
+    "add the document to the index by specifying doc_id, doc_text, and doc_embeddings. Please note: if you are embedding all the entries in the Amazon product catalog dataset, this could take several minutes."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "3df2a7e1",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "for doc_id, doc_text in dataset:\n",
+    "    embedding = embedding_model.encodeDocs([doc_text])\n",
+    "    our_index.add_doc(doc_id=doc_id, doc_text=doc_text, doc_embeddings=embedding)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "2d422245",
+   "metadata": {},
+   "source": [
+    "# Step 4: Querying the index\n",
+    "\n",
+    "The following cell starts up an interactive demo, where you can type any query you want and the index will return the document most semantically likely to answer your query. \n",
+    "\n",
+    "If you are using the very small example dataset we prepopulated above, try queries like \"what is deep learning\" or \"how do auto encoders work\".\n"
```

Comment:
Can you add example queries for the product notebook? Ask anshu I think he has some